File: ios/Runner/Bridge/VpnEventListener.swift
import Foundation

final class VpnEventListener: VpnEventsStreamHandler {
  static let shared = VpnEventListener()

  private let stateLock = NSLock()
  private var eventSink: PigeonEventSink<VpnEventMessage>?
  private var storedStatus = StatusChangedMessage(status: .disconnected)
  private var rxBytes: Int64 = 0
  private var txBytes: Int64 = 0

  var lastStatus: StatusChangedMessage {
    stateLock.lock()
    defer { stateLock.unlock() }
    return storedStatus
  }

  override func onListen(withArguments arguments: Any?, sink: PigeonEventSink<VpnEventMessage>) {
    eventSink = sink
    emit(lastStatus)
  }

  override func onCancel(withArguments arguments: Any?) {
    eventSink = nil
  }

  func emit(_ event: VpnEventMessage) {
    stateLock.lock()
    if let status = event as? StatusChangedMessage {
      storedStatus = status
    } else if let traffic = event as? TrafficChangedMessage {
      rxBytes = traffic.rxBytes
      txBytes = traffic.txBytes
    }
    stateLock.unlock()
    DispatchQueue.main.async { [weak self] in
      self?.eventSink?.success(event)
    }
  }

  func snapshot() -> VpnStatusSnapshotMessage {
    stateLock.lock()
    defer { stateLock.unlock() }
    return VpnStatusSnapshotMessage(
      status: storedStatus.status,
      connectedSinceEpochMs: storedStatus.connectedSinceEpochMs,
      rxBytes: rxBytes,
      txBytes: txBytes
    )
  }
}

File: android/app/src/main/kotlin/com/example/vpn_osin/bridge/VpnEventListener.kt
package com.example.vpn_osin.bridge

import android.os.Handler
import android.os.Looper

class VpnEventListener : VpnEventsStreamHandler() {
    private val mainHandler = Handler(Looper.getMainLooper())
    private var sink: PigeonEventSink<VpnEventMessage>? = null
    private val forward: (VpnEventMessage) -> Unit = { event ->
        mainHandler.post { sink?.success(event) }
    }

    override fun onListen(p0: Any?, sink: PigeonEventSink<VpnEventMessage>) {
        this.sink = sink
        VpnEventBus.addListener(forward)
    }

    override fun onCancel(p0: Any?) {
        VpnEventBus.removeListener(forward)
        sink = null
    }
}
